### PR TITLE
Use symlinks instead of copies for test result_images.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

This seems to shave a few minutes(!) from the CI.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by tacaswell** [comment]
power-cycled to get the gdb fix.

**Comment by jklymak** [comment]
OK, but are these tests really any faster on CI?  15 minutes 37 sec for the 3.6 test doesn't seem any faster than other recent tests...  https://travis-ci.org/matplotlib/matplotlib/builds/325962079?utm_source=github_status&utm_medium=notification took 14 minutes 50 seconds (as a completely non-scientific comparison).

**Comment by anntzer** [comment]
Seems quite variable.  Will probably try to accumulate some statistics on my side first...

File: lib/matplotlib/testing/decorators.py [hunk 1]
```diff
@@ -192,13 +192,19 @@ def copy_baseline(self, baseline, extension):
...
             'expected')
-        if os.path.exists(orig_expected_fname):
-            shutil.copyfile(orig_expected_fname, expected_fname)
-        else:
-            reason = ("Do not have baseline image {} because this "
-                      "file does not exist: {}".format(expected_fname,
-                                                       orig_expected_fname))
-            raise ImageComparisonFailure(reason)
+        try:
+            try:  # os.symlink errors if the target already exists.
```
Comment (timhoffm) [inline lib/matplotlib/testing/decorators.py]: Better:
~~~
with contextlib.suppress(FileNotFoundError):
    os.remove(expected_fname)
~~~

**Comment by timhoffm** [comment]
What is the right way forward here? Should we try to verify the speed gain again? Should we drop this? Should we just accept on the basis that symlinking feels better and may even bring a performance gain (without test) and it's just slightly more complicated.

**Comment by WeatherGod** [comment]
Are there going to be issues on Windows filesystems?

**Comment by timhoffm** [comment]
You need privileges to symlink on windows, which a regular user on Windows does not have (https://docs.python.org/3/library/os.html?highlight=symlink#os.symlink). This is covered in the code with the `except OSError` and copied instead.

The code is not compatible with WindowsXP and earlier, as that would throw a `NotImplementedError`. Probably we don't have to worry about testing for XP, but one could check that as well and resort to copying again.

File: lib/matplotlib/testing/decorators.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/testing/decorators.py]: sure

**Comment by anntzer** [comment]
It feels better to use symlinks, although I have admittedly no numbers to show real improvements.  I don't really care if this goes in or if we abandon it.
The OSError check will handle Windows just fine.

**Pull request merged**